thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/HoltWintersDetector.java

```diff
@@ -0,0 +1,538 @@
+
+    // anomalies
+    List<MergedAnomalyResultDTO> result = DetectionUtils.makeAnomalies(sliceData, df, COL_ANOMALY,
+        window.getEndMillis(),
+        DetectionUtils.getMonitoringGranularityPeriod(monitoringGranularity, datasetConfig), datasetConfig);
+
+    return result;
+  }
+
+  /**
+   * Break a time period into parts and fetch it one by one, and put it together into a DataFrame
+   *
+   * @param metricEntity metric entity
+   * @param start start timestamp
+   * @param end end timestamp
+   * @return Data Frame that has data from start to end
+   */
+  private DataFrame fetchDataByPart(MetricEntity metricEntity, long start, long end) {
+    long duration = end - start;
+    List<MetricSlice> slices = new ArrayList<>();
+    long temp = start;
+    int parts = (int) (duration / PARTITION_PERIOD);
+    for (int i = 0; i < parts; i++) {
+      MetricSlice sliceData = MetricSlice.from(metricEntity.getId(), temp, temp + PARTITION_PERIOD,
+          metricEntity.getFilters(), timeGranularity);
+      temp += PARTITION_PERIOD;
+      slices.add(sliceData);
+    }
+    MetricSlice sliceData = MetricSlice.from(metricEntity.getId(), temp, end,
+        metricEntity.getFilters(), timeGranularity);
+    slices.add(sliceData);
+
+    InputData data = this.dataFetcher.fetchData(new InputDataSpec().withTimeseriesSlices(slices)
+        .withMetricIdsForDataset(Collections.singletonList(metricEntity.getId())));
+
+    LongSeries longSeries = LongSeries.buildFrom();
+    DoubleSeries doubleSeries = DoubleSeries.buildFrom();
+
+    DataFrame df = new DataFrame();
+    df.addSeries(COL_TIME, longSeries);
+    df.addSeries(COL_VALUE, doubleSeries);
+    df.setIndex(COL_TIME);
+
+    Set<Map.Entry<MetricSlice, DataFrame>> set = data.getTimeseries().entrySet();
+    for (Map.Entry<MetricSlice, DataFrame> entry: set) {
+      df = df.append(entry.getValue());
+    }
+    df = df.sortedBy(COL_TIME);
+
+    // Remove duplicates
+    long prev = -1;
+    for (int i = 0; i < df.size(); i++) {
+      long curr = df.getLongs(COL_TIME).get(i);
+      if (curr == prev) {
+        DataFrame df1 = df.slice(0, i);
+        DataFrame df2 = df.slice(i+1, df.size());
+        df = df1.append(df2);
+      }
+      prev = curr;
+    }
+    return df;
+  }
+
+  /**
+   * Returns a data frame containing the same time daily data
+   * @param originalDF the original dataframe
+   * @param time the time of day
+   * @return dataframe containing same time of daily data for LOOKBACK number of days
+   */
+  private DataFrame getDailyDF(DataFrame originalDF, Long time) {
+    LongSeries longSeries = (LongSeries) originalDF.get(COL_TIME);
+    Long start = longSeries.getLong(0);
+    DateTime dt = new DateTime(time);
+    DataFrame df = DataFrame.builder(COL_TIME, COL_VALUE).build();
+
+    for (int i = 0; i < LOOKBACK; i++) {
+      DateTime subDt = new DateTime(dt);
+      subDt = subDt.minusDays(1);
+      long t = subDt.getMillis();
```

Comment:
For l248 to l250 you can use 
long t = dt.minusDays(1).getMills();

Reply:
I had to do subDt = subDt.minusDays(1), because subDt.minusDays(1) does not mutate subDt

Reply:
I need to do subDt = subDt.minusDays(1); because I will need subDt to be changed for the code for getting week over week data below.

Reply:
Then why not DateTime subDt  = dt.minusDays(1) directly?